fix hops seed drop and add recipe

Hojosa opened the pull request:
closed point 1 and 3 from #185

Hojosa committed:
drop grape seeds correctly

Hojosa committed:
add recipe for hops seeds

moonfather1 commented:
confused,  first of all, loot table seems weird with repeated block. maybe a bad paste?

i can't read that, i'll try in-game.
here's how i think the loot table should work: bottom-most block should drop 0-1 seeds, hanging ones zero. max-stage blocks should drop hops when destroyed. there is also a matter of dropping rope (if not bottom-most), but sure about chance.

Hojosa commented:
it looks like it but no, its correct.
the first block makes sure that a seed is dropped at AGE:1 (this was the faulty block, i would say)
second block drops hops on age 7, and the last one drops extra hops on age 7 (i think its 1-2 or 1-3, not sure)

moonfather1 commented:
don't bother, it's too complicated for me to look at.   i'll just paste it into my dev env and see.

Alatyami commented:
I see what you did. Since there is a recipe for Hops to Seeds, then if AGE == 1, drop seeds, AGE 2 through 6 drop nothing, AGE 7 drop hops. Not sure I like that. If you accidentally destroy your crop before AGE 7 then you have to go find seeds again.

Should be similar to other crops as they drop the seed if not mature. And since you do not destroy the hops crop to get the hops, player will not get extra seeds as they harvest.

Hojosa commented:
actually, in my testing, all ages are dropping 1 seed now, when the plant is destroyed. not sure why, it looks to me, as if its an inclusive or <= check.

moonfather1 commented:
ok, i've overwritten the hops_vine.json in jar file with the one from this PR. here's what i determined:

* non-max-age crop block drops 1 seed always.
* max-age block drops a few hops.
* no rope is dropped.

my suggestions: 
* level 2 (two blocks above farmland) and level 3 shouldn't drop seeds, although it is not a disaster if they do.
* level 2 and 3 shouldn't drop rope - a rope block should be created in world when a hops block is destroyed.

unrelated:
* destroying level 2 hops crop leaves level 3 above it. i should make a separate issue for this.

Hojosa committed:
ensure crops check, if they can survive on the given pos and place rope on removal, when they are not the root plant

Hojosa commented:
this commit fixes #207
not only do the crops now check, if they can actually survive at the given pos. (no floating crops anymore)
they also place back rope, if they are not the root plant. 
this include changes to the vine leaves as well, as they are also affected by this. (same base class)

about hops lv 2 and 3 shouldnt drop seeds. i would leave it as it is for now. i dont know if or how the loottable could handle this.

moonfather1 commented:
rope is now preserved if breaking from top downwards but lost if breaking from the bottom. when checking whether  `below.is(this)`  maybe allow  `isAir()` ... it'll be false if below is farmland but true if block below was killed by  `canSurvive`  returning false. maybe also allow below being rope.

moonfather1 commented:
>  i would leave it as it is for now. i dont know if or how the loottable could handle this.

not worth the time, that is such a minutiae compared to actual issues we allowed this block to have.

Hojosa committed:
modify canSurive, to account for block removal

Hojosa committed:
modify hops lootTable to drop seeds corretly

Hojosa committed:
add lootTable for grap vines

Hojosa commented:
so i think whats happening is this:
when you remove a block starting below, the blocks that are left will be removed, because canSurive return false. 
So i modified this method to replace our crops block with rope instead and return true. 

odd thing now, if you remove the root plant (the one on the farmland) the blocks above are not removed right away. dont know why. 

and i added the missing loot table for the grape vines. noticed that you just lost your seed if you removed the plant. 
also tweaked the hops table, should now drop seeds more corretly.

moonfather1 commented:
new change to `canSurvive` can and will cause problems. it should not have side effects.

Hojosa committed:
revert canSurvive changes and implement replacement logic in onRemove

Hojosa commented:
good point. i looked at it again and i overlooked that onRemove also gets called, when the block is removed via onSurvive returning false. 
i reverted the changes to onSurvive and added the check to onRemove. this works and is much cleaner.

moonfather1 commented:
returning rope can be much simpler - you can just check the block below. nothing else. if below is farmland (preferably tag) then there was no rope there, do nothing; if anything but farmland, then there was surely rope there before hops grew into the block.
and maybe allow dirt tag instead of farmland to account for farmland trampled in meantime.

as for canSurvive(), i don't understand the second part. maybe needed for worldgen?

Hojosa commented:
canSurvive(), the second part is for the upper blocks. because any block on level 2 or higher will never have farmland below it. 

for returning rope, you are right. sadly there is no tag for farmland, so i used the canSustainPlant check, which does exactly that. also work with farmers delight farmland.

Hojosa committed:
further improvements to placing rope back, on block removal

moonfather1 commented:
good thing is - hops work just fine.

@Hojosa 
bad thing is that i had to confirm in-game that things work, and then read the code vary carefully because i expected one thing not to work. so - partly my bad. in `canSurvive()`, line:
`return pLevel.getBlockState(pPos.below()).canSustainPlant(pLevel, blockpos, Direction.UP, this) || pLevel.getBlockState(blockpos).is(this);`
confused me because i saw `somePos.below()` and then later `anotherBlockPos` twice and in casual reading i understood that former is the position below current (correct) and that later is current position (wrong). confusing code, but technically correct.

Alatyami committed:
Merge branch 'development' into 185-Hops_are_weird
